RealmSwift/SwiftUI.swift

```diff
@@ -643,6 +647,368 @@ extension EnvironmentValues {
+        self.partitionValue = partitionValue
+    }
+}
+
+/**
+An enum representing different states from `AsyncOpen` and `AutoOpen` process
+*/
+public enum AsyncOpenState {
+    /// Starting the Realm.asyncOpen process.
+    case connecting
+    /// Waiting for a user to be logged in before executing Realm.asyncOpen.
+    case waitingForUser
+    /// The Realm has been opened and is ready for use. For AsyncOpen this means that the Realm has been fully downloaded, but for AutoOpen the existing local file may have been used if the device is offline.
+    case open(Realm)
+    /// The Realm is currently being downloaded from the server.
+    case progress(Progress)
+    /// Opening the Realm failed.
+    case error(Error)
+}
+
+// MARK: - AsyncOpen
+
+/// A property wrapper type that initiates a `Realm.asyncOpen()` for the current user which asynchronously open a Realm,
+/// and notifies states for the given process
+///
+/// Add AsyncOpen to your ``SwiftUI/View`` or ``SwiftUI/App``,  after a user is already logged in,
+/// or if a user is going to be logged in
+///
+///     @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen
+///
+/// This will immediately initiates a `Realm.asyncOpen()` operation which will perform all work needed to get the Realm to
+/// a usable state. (see Realm.asyncOpen() documentation)
+///
+/// This property wrapper will publish states of the current `Realm.asyncOpen()` process like progress, errors and an opened realm,
+/// which can be used to update the view
+///
+///     struct AsyncOpenView: View {
+///         @AsyncOpen(appId: "app_id", partitionValue: <partition_value>) var asyncOpen
+///
+///         var body: some View {
+///            switch asyncOpen {
+///            case .notOpen:
+///                ProgressView()
+///            case .open(let realm):
+///                ListView()
+///                   .environment(\.realm, realm)
+///            case .error(_):
+///                ErrorView()
+///            case .progress(let progress):
+///                ProgressView(progress)
+///            }
+///         }
+///     }
+///
+/// This opened `realm` can be later injected to the view as an environment value which will be used by our property wrappers
+/// to populate the view with data from the opened realm
+///
+///     ListView()
+///        .environment(\.realm, realm)
+///
+@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
+@propertyWrapper public struct AsyncOpen<Partition>: DynamicProperty where Partition: BSON {
+    @Environment(\.realmConfiguration) var configuration
+    @Environment(\.partitionValue) var partitionValue
+    @ObservedObject private var storage: ObservableAsyncOpenStorage
+
+    private func asyncOpen() {
+        storage.asyncOpen()
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+                cancel()
```

Comment:
Why does this cancel even though it's completed?

Reply:
you are right, this shouldn't be done after completed